File: 2022/day08.py
import time
from itertools import count

start = time.time()
with open("../inputs/2022/day08.input") as f:
    data = f.read()
    inp = [list(map(int, line)) for line in data.split("\n")]


def part1(inp: list) -> int:
    visible = set()
    for i, line in enumerate(inp):
        maxseen = -1
        for j, c in enumerate(line):
            if c > maxseen:
                visible.add((i, j))
                maxseen = c
        maxseen = -1
        for j, c in enumerate(line[::-1]):
            if c > maxseen:
                visible.add((i, (len(line) - j - 1)))
                maxseen = c
    transposed = list(zip(*inp))
    for j, c in enumerate(transposed):
        rowmax = -1
        for i, line in enumerate(c):
            if line > rowmax:
                visible.add((i, j))
                rowmax = line
        rowmax = -1
        for i, line in enumerate(c[::-1]):
            if line > rowmax:
                visible.add((len(c) - i - 1, j))
                rowmax = line
    return len(visible)


def ray_length(line, pos, direction, n):
    val = line[pos]
    for i in count():
        pos += direction
        if pos == -1 or pos == n:
            return i
        if line[pos] >= val:
            return i + 1
    raise ValueError


def score(i, j, row, col, n):
    return (
        ray_length(row, j, -1, n)
        * ray_length(row, j, 1, n)
        * ray_length(col, i, -1, n)
        * ray_length(col, i, 1, n)
    )


def part2(inp: list) -> int:
    n = len(inp)
    k = len(inp[0])
    transposed = list(zip(*inp))
    return max(
        score(i, j, inp[i], transposed[j], n) for i in range(n) for j in range(k)
    )


print(
    f"Day 08     {part1(inp):<14} {part2(inp):<14} {(time.time() - start)*1e3:>11.2f}"
)

File: rust-2022/src/day08.rs
use itertools::Itertools;

const P: usize = 99;
pub fn read(inp: &[u8]) -> Vec<u8> {
    let mut v = vec![0_u8; P * P];
    let mut i = 0;
    for &c in inp {
        if c != b'\n' {
            v[i] = c;
            i += 1;
        }
    }
    v
}

pub fn part1(inp: &[u8]) -> u16 {
    let mut visible = vec![0_u8; inp.len()];
    for row in 0..P {
        let mut max = 0;
        for col in 0..P {
            let c = inp[row * P + col];
            if c > max {
                max = c;
                visible[row * P + col] = 1;
            }
        }
        max = 0;
        for col in (0..P).rev() {
            let c = inp[row * P + col];
            if c > max {
                max = c;
                visible[row * P + col] = 1;
            }
        }
    }
    for col in 0..P {
        let mut max = 0;
        for row in 0..P {
            let c = inp[row * P + col];
            if c > max {
                max = c;
                visible[row * P + col] = 1;
            }
        }
        max = 0;
        for row in (0..P).rev() {
            let c = inp[row * P + col];
            if c > max {
                max = c;
                visible[row * P + col] = 1;
            }
        }
    }
    visible.iter().map(|&x| x as u16).sum::<u16>()
}

pub fn part2(inp: &[u8]) -> u32 {
    let ray = |mut i, mut j, (istride, jstride)| -> u32 {
        let val = inp[i * P + j];
        for c in 0.. {
            i = (i as i64 + istride) as usize;
            j = (j as i64 + jstride) as usize;
            if i >= P || j >= P {
                return c;
            }
            if inp[i * P + j] >= val {
                return c + 1;
            }
        }
        100
    };
    (0..P).cartesian_product(0..P).fold(0u32, |acc, (i, j)| {
        acc.max(
            [(-1, 0), (1, 0), (0, -1), (0, 1)]
                .iter()
                .fold(1u32, |acc, &x| acc * ray(i, j, x)),
        )
    })
}

pub fn run() -> (String, String) {
    let inp = include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day08.input"));

    let data = read(inp);
    let a = part1(&data).to_string();
    let b = part2(&data).to_string();
    (a, b)
}
